runtime/marshal_httpbodyproto.go

```diff
@@ -0,0 +1,61 @@
+package runtime
+
+import (
+	"io"
+
+	"google.golang.org/genproto/googleapis/api/httpbody"
+)
+
+// SetHTTPBodyMarshaler overwrite  the default marshaler with the HTTPBodyMarshaler
+func SetHTTPBodyMarshaler(serveMux *ServeMux) {
+	serveMux.marshalers.mimeMap[MIMEWildcard] = &HTTPBodyMarshaler{
+		&JSONPb{OrigName: true},
+	}
+}
+
+// HTTPBodyMarshaler is a Marshaler which supports marshaling of a
+// google.api.HttpBody message as the full response body if it is
+// the actual message used as the response. If not, then this will
+// simply fallback to the JSONPb marshaler.
+type HTTPBodyMarshaler struct {
+	Marshaler
+}
+
+// ContentType implementation to keep backwards compatability with marshal interface
+func (h *HTTPBodyMarshaler) ContentType() string {
+	return h.ContentTypeFromMessage(nil)
+}
+
+// ContentTypeFromMessage in case v is a google.api.HttpBody message it returns
+// its specified content type otherwise fall back to the JSONPb marshaler.
+func (h *HTTPBodyMarshaler) ContentTypeFromMessage(v interface{}) string {
+	if httpBody, ok := v.(*httpbody.HttpBody); ok {
+		return httpBody.GetContentType()
+	}
+	return h.Marshaler.ContentType()
+}
+
+// Marshal marshals "v" by returning the body bytes if v is a
+// google.api.HttpBody message, otherwise it falls back to the JSONPb marshaler.
+func (h *HTTPBodyMarshaler) Marshal(v interface{}) ([]byte, error) {
+	if httpBody, ok := v.(*httpbody.HttpBody); ok {
+		return httpBody.Data, nil
+	}
+	return h.Marshaler.Marshal(v)
+}
+
+// Unmarshal unmarshals JSON data into "v".
+// google.api.HttpBody messages are not supported for request messages.
+func (h *HTTPBodyMarshaler) Unmarshal(data []byte, v interface{}) error {
+	return h.Marshaler.Unmarshal(data, v)
+}
+
+// NewDecoder returns a Decoder which reads JSON stream from "r".
+func (h *HTTPBodyMarshaler) NewDecoder(r io.Reader) Decoder {
+	return h.Marshaler.NewDecoder(r)
+}
+
+// NewEncoder returns an Encoder which writes JSON stream into "w".
+func (h *HTTPBodyMarshaler) NewEncoder(w io.Writer) Encoder {
+	return h.Marshaler.NewEncoder(w)
+}
```

Comment:
You can remove `NewEncoder`, `NewDecoder` and `Unmarshal` now. The embedded methods will be called automatically.

Reply:
Good point :)